piptools/utils.py

```diff
@@ -121,6 +122,9 @@ def format_requirement(
     elif is_url_requirement(ireq):
         line = _build_direct_reference_best_efforts(ireq)
     else:
+        # Canonicalize the requirement name
+        # https://packaging.pypa.io/en/latest/utils.html#packaging.utils.canonicalize_name
+        ireq.req.name = canonicalize_name(ireq.req.name)
```

Comment:
🤔 this is a formatting-only helper, but here it silently mutates an attribute of an attribute of a passed-in object. this looks suspicious since it mutates state that seemingly is not owned by this function; perhaps there is a way this can be done in a ‘pure’ / immutable / read-only way?

Reply:
seems `copy.copy()` would do the trick:

```suggestion
        req = copy.copy(ireq.req)
        req.name = canonicalize_name(req.name)
        line = str(req)
```